### PR TITLE
ARROW-12575: [R] Use unary negative kernel

### PR DESCRIPTION
**Pull request by thisisnic**

### CODE REVIEW

File: r/R/expression.R [hunk 1]
```diff
@@ -56,13 +56,9 @@ build_array_expression <- function(FUN,
                                    args = list(...),
                                    options = empty_named_list()) {
   if (FUN == "-" && length(args) == 1L) {
-    # Unary -, i.e. just make it negative, and somehow this works
     if (inherits(args[[1]], c("ArrowObject", "array_expression"))) {
-      # Make it be 0 - arg
-      # TODO(ARROW-11950): do this in C++ compute
-      args <- list(0L, args[[1]])
+      return(build_array_expression("negate", args[[1]]))
```
Comment (nealrichardson) [inline r/R/expression.R:60]: Question: do we want to use "negate" or "negate_checked" here? https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_arithmetic.cc#L514-L524

We have tended to use the "checked" versions of the other arithmetic functions, but we haven't thought too hard about it.

Comment (bkietz) [reply r/R/expression.R:60]: I think the edge case which `negate_checked` watches for is probably not necessary here. @edponce

Comment (thisisnic) [reply r/R/expression.R]: Perhaps for the sake of consistency, using `negate_checked` makes sense, even if not strictly necessary?

Comment (bkietz) [reply r/R/expression.R]: Alright, note that `negate_checked` does not support unsigned arguments:
```r
call_function("negate_checked", args=list(Array$create(c(1,2), type=uint8())))
#> Error: NotImplemented: Function negate_checked has no kernel matching input types (array[uint8])
```

**Comment by ianmcook** [comment]
@thisisnic could you add some tests with multiple unary negative operators stacked up. Maybe test with odd & even numbers of them, and parens also, e.g.
```
---2
-(-(-2))
--2
-(-(2))

File: r/R/expression.R (hunk 1 above)
Comment (edponce) [reply r/R/expression.R]: I just noticed that we should update the *negate_checked_doc* to state explicitly that it does not support unsigned arguments. We do show this in the function signature table https://github.com/apache/arrow/blob/master/docs/source/cpp/compute.rst#arithmetic-functions

**Comment by nealrichardson** [comment]
> @thisisnic could you add some tests with multiple unary negative operators stacked up. Maybe test with odd & even numbers of them, and parens also, e.g.
> 
> ```
> ---2
> -(-(-2))
> --2
> -(-(2))
> ```

Wouldn't this just be testing R's parser?

File: r/R/expression.R (hunk 1 above)
Comment (edponce) [reply r/R/expression.R]: I would think that using the *checked* versions for language bindings is more desirable as you want errors to be triggered ASAP. The *unchecked* versions should be used knowing the wrap around behavior may occur. I think of it as *unchecked = safe math* and *checked = true math*.

Comment (edponce) [reply r/R/expression.R]: I just noticed that we should update the negate_checked_doc to state explicitly that it does not support unsigned arguments. We do show this in the function signature table https://github.com/apache/arrow/blob/master/docs/source/cpp/compute.rst#arithmetic-functions

Comment (edponce) [reply r/R/expression.R]: I would think that using the checked versions for language bindings is more desirable as you want errors to be triggered ASAP. The unchecked versions should be used knowing the wrap around behavior may occur. I think of it as unchecked = safe math and checked = true math

**Comment by edponce** [comment]
The multiple negations and parenthesis are resolved on R side as only the resolved value is passed to the C++ backend. If the expression was passed as a string then it would make sense to check these cases.

When we did the C++ tests, we checked for nested negations and they passed (although these tests were not included in the final PR).

**Comment by ianmcook** [comment]
> Wouldn't this just be testing R's parser?

In theory yes? I tend to have this philosophy re tests: https://twitter.com/gshotwell/status/1389563734679605251?s=20

**Review by nealrichardson — APPROVED** [review]

**Pull request closed without merging**